consensus/spos/bls/subroundBlock.go

```diff
@@ -438,8 +494,102 @@ func (sr *subroundBlock) receivedBlockBodyAndHeader(ctx context.Context, cnsDta
 	return blockProcessedWithSuccess
 }
 
+func (sr *subroundBlock) verifyProof() bool {
+	previousAggregatedSignature, previousBitmap := sr.Header.GetPreviousAggregatedSignatureAndBitmap()
+	hasProof := len(previousAggregatedSignature) > 0 && len(previousBitmap) > 0
+	hasLeaderSignature := len(sr.Header.GetLeaderSignature()) != 0
+	isFlagEnabled := sr.EnableEpochsHandler().IsFlagEnabled(common.ConsensusPropagationChangesFlag)
+	if isFlagEnabled && !hasProof {
+		log.Debug("received header without proof after flag activation")
+		return false
+	}
+	if !isFlagEnabled && hasProof {
+		log.Debug("received header with proof before flag activation")
+		return false
+	}
+	if isFlagEnabled && hasLeaderSignature {
+		log.Debug("received header with leader signature after flag activation")
+		return false
+	}
+
+	return true
+}
+
+func (sr *subroundBlock) saveLeaderSignature(nodeKey []byte, signature []byte) error {
+	if !sr.EnableEpochsHandler().IsFlagEnabled(common.ConsensusPropagationChangesFlag) {
+		return nil
+	}
+
+	if len(signature) == 0 {
+		return spos.ErrNilSignature
+	}
+
+	if sr.IsSelfLeaderInCurrentRound() || sr.IsMultiKeyLeaderInCurrentRound() {
```

Comment:
why not save the sig share if the node is leader as well?

we would need it for aggregation afterwards if the node is leader.

Reply:
right, updated